# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [0]:
import math

class DualNumber:
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue

    def __str__(self):
        return str(self.value) + " + " + str(self.dvalue) + "ε"

    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    
    def __add__(self, other):
      return DualNumber(self.value + other.value, self.dvalue + other.dvalue)
    
    def __truediv__(self, other):
      return DualNumber(self.value/other.value,
          (self.dvalue*other.value - self.value*other.dvalue)/(other.value**2))
      
    def __sub__(self, other):
      return DualNumber(self.value - other.value, self.dvalue - other.dvalue)
    
    def __pow__(self, other):
      return DualNumber(self.value**other.value, (self.value**other.value) * (other.dvalue * math.log1p(self.value) + self.dvalue*other.value/self.value)) 

In [19]:
# Tests

DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0)


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [0]:
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [0]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    return DualNumber(math.cos(x.value), -math.sin(x.value)*x.dvalue)

def tan(x):
   return sin(x)/cos(x)

def exp(x):
    return DualNumber(math.exp(x.value), math.exp(x.value)*x.dvalue)

In [0]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [23]:
def z(x,y):
  return x* y + math.sin(x)

def zDerivativeX(x,y):
  return y + math.cos(x)

def zDerivativeY(x,y):
  return x

x = 0.5
y = 4.2

result = DualNumber(z(x,y), zDerivativeX(x,y))

print(result)

2.579425538604203 + 5.077582561890373ε


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

$z = x \cdot y + sin(x) $

$\frac{\partial z}{\partial x}=y + cos(x)$

$z = 0.5 \cdot 4.2 + sin(0.5) = 2.579425539$

$z' = 4.2 + cos(0.5) = 5.077582562$

$result = z + z'\epsilon = 2.579425539 + 5.077582562\epsilon$

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [24]:
dzdy = zDerivativeY(x,y)

print('dz/dy:', dzdy)

dz/dy: 0.5


In [0]:
#Tests
assert dzdy
assert type(dzdy) == float


__Task:__ Finally, use the code block below to experiment and test the other math functions and methods you created.

In [26]:
addition = result + result
print(addition)
subtraction = result - result
print(subtraction)
multiplication = result * result
print(multiplication)
division = result/result
print(division)
print(sin(result))
print(cos(result))
print(tan(result))
print(exp(result))
answer = result**result
print(answer)

5.158851077208406 + 10.155165123780746ε
0.0 + 0.0ε
6.653436109203583 + 26.19449226902277ε
1.0 + 0.0ε
0.5330210484038216 + -4.296152663425431ε
-0.8461019808264787 + -2.7064583804957687ε
-0.6299725807084894 + 7.092699662639372ε
13.189559090400415 + 66.9710752364398ε
11.521072885937572 + 133.0975129845776ε
